### DISCUSSION TITLE
Add a custom debug device for analysing the framebuffer contents

### DISCUSSION THREAD

**Discussion by iamgreaser** [86Box/86Box#4004] [Ideas]
**SPECIFICATION BELOW IS OUT OF DATE** - please see `doc/specifications/86box-unit-tester.md` where present or pull request #4011 if it's not merged yet. It may still be accurate to what's implemented, but I haven't checked.

We have a surprising amount of stuff which uses the SVGA renderer, and it's extremely difficult to not break headcanonically-correct behaviour in chipsets separate from the ones you're working on.

Now, the correct approach would involve a good balance between reusability and customisability and it's something I haven't fully got my head around yet. But in the meantime, I do need a way to confirm that, say, VGA hasn't broken because some SVGA chipset someone is working on emulating appears to be doing things weirdly.

Here are my thoughts on how this could work combined with a proposal which could work quite well.

----

Updates:

Times are very approximate, and in UTC.

* 2024-01-07T01:35:
  - Added "monitor" field to command 0x01 ("capture snapshot").
  - Defined the behaviour of the initial state of the snapshot buffer.
  - Explicitly state that everything is little-endian.

* 2024-01-07T00:36:
  - Changed capitalisation of "86BOX" activation sequence to "86Box".
  
* 2024-01-06T22:54:
  - Added requirement 8 (produce result and exit)
  - Added a way to close 86Box with the given exit code (command 0x04).

----

Requirements:

1. There will be a way to analyse the framebuffer quickly.
2. There will be a way to analyse the framebuffer thoroughly. This also means that one must be able to capture a snapshot of the framebuffer for analysis.
3. We only need one framebuffer capture.
4. Framebuffer captures should be doable at pretty much any point.
5. It will work on all boards without conflicts (within reason - none of this "let's use port 0x80 for something important because we need to break Linux" scumbaggery).
6. One must be able to disable this device.
7. Ideally, this would be extensible and reusable for testing other things.
8. Ideally, this would be able to produce a result for an external program and exit without user input. Makes it easier to make 86Box actually run under CI, and also enables us to leverage the truly glorious `git bisect run` (if you've never used it, you have no idea how amazing and brutally effective it is at hunting down regressions).

Of course, the K.I.S.S. principle applies (Keep It Simple, Stupid) - if it's too complex, it's probably not going to be implemented or used.

Non-requirements:

- We are not attempting to emulate something realistically in terms of timing. When building or analysing a snapshot, one is allowed to lag the hell out of the emulator if necessary. (But please don't.)

----

Proposal:

An 8-bit I/O device which can be placed pretty much anywhere in I/O space, and initially is invisible. There are 2 adjacent I/O addresses used, the first for command(W)/status(R), the second for data.

A base I/O address is set via a sequence of byte writes on port 0x80. I'm going to exercise a lack of imagination and say that if the byte sequence of `'8', '6', 'B', 'o', 'x'` is sent to port 0x80, then the next 2 bytes written will be the LSbyte and MSbyte of the I/O address. If an address of `0xFFFF` is chosen then the device is unavailable. (requirement 5) (requirement 6)

The port used to actually activate this will (eventually) need to be configurable as, for example, the PC-98 (not IBM PC compatible at all) has a 5.25" floppy disk controller at that port. But for now, 0x80 is the port to snoop... and also we need to ensure that the POST card doesn't break when we have this enabled. (requirement 5)

The snapshot framebuffer is 8:8:8:8 B:G:R:X, left-to-right then top-to-bottom. The rightmost pixel of row y is followed by the leftmost pixel of row y+1. After each B,G,R value, there is a 0x00 byte, unless a value is outside of the overscan area in which case R=G=B=0x00 and X=0xFF.

Prior to any snapshot being taken, the snapshot is to be treated as a 0x0 image with a 0x0 overscan and an image offset of (0, 0).

Coordinates are such that (0,0) denotes the top-left pixel of the non-overscanned image, and negative coordinates denote stuff in the left and upper overscan area.

Framebuffer data can either be read directly (requirement 2), or a CRC-32 can be taken of the data (requirement 1). The CRC-32 is as per zlib (Galois LFSR 0xEDB88320 shifting right, initial seed 0xFFFFFFFF, final value XORed with 0xFFFFFFFF).

All values are little-endian.

W BASE+0x00 = command - writing to this aborts any previous commands.
* 0x00 = clear previous command (essentially a no-op)
* 0x01 = capture snapshot (requirement 4) (requirement 3)
  - W uint8_t monitor
    - 0x00 = no monitor (that is, clear the snapshot)
    - 0x01 = primary monitor
    - 0x02 = secondary monitor
    - Any monitor which doesn't exist will be treated as having all dimensions and offsets as 0, and the snapshot buffer will be cleared.
  - R uint16_t image width in pixels
  - R uint16_t image height in pixels
  - R uint16_t overscanned width in pixels
  - R uint16_t overscanned height in pixels
  - R uint16_t X start of image from overscan
  - R uint16_t Y start of image from overscan
* 0x02 = read snapshot rectangle (requirement 2)
  - W uint16_t selection width in pixels
  - W uint16_t selection height in pixels
  - W int16_t selection x offset
  - W int16_t selection y offset
  - R uint8_t[height][width][4] pixels
* 0x03 = read snapshot rectangle CRC-32 (requirement 1)
  - W uint16_t selection width in pixels
  - W uint16_t selection height in pixels
  - W int16_t selection x offset
  - W int16_t selection y offset
  - R uint32_t crc
* 0x04 = exit 86Box with the given exit code (requirement 8)
  - W = exit code - the actual exit code will be clamped to the highest possible value which tends to be 127.

R BASE+0x00 = status - must not be 0xFF, ever.
* bit 0: there is data to be read
* bit 1: one needs to send more data for the given command
* bit 2: there is no command in flight
* bit 3: the supplied command does not exist (requirement 7)
* remaining bits are 0.

W BASE+0x01 = data write

R BASE+0x01 = data read - 0xFF if nothing to be read (one should check status register bit 0 instead of relying on this)

----

Thoughts?

**Comment by dhrdlicka** [comment]
Considering that this is a "device" intended purely intended for our debugging purposes, is there really any point in the handshake sequence? Can't we just set the base I/O port in the config?

**Reply by iamgreaser** [reply]
I tend to have a soft spot for autodetection.

Now, arguably, for maximum compatibility, ISA-PnP is the correct answer -- wait, no it isn't, ISA-PnP is never the answer it's a horrible thing to actually use please never ever suggest ISA-PnP.

But basically, you can either require each unit testing program to configure a port, or you can give it some ability to select a port that you know might not have something else in the same place.

Having said *that*, there may be cases where we need to place the handshake port elsewhere (I've been referring to this as the "trigger" port but "handshake" really is a better word). But I think that's mostly going to be for non-IBM-PC stuff like PC-98.
